geoips/pydantic/products.py

```diff
@@ -0,0 +1,231 @@
+        # Delegate arguments validation to each plugin type argument class
+        plugin_type = values.get("type", " ").lower()
+        plugin_type_camel_case = "".join(
+            [word.capitalize() for word in plugin_type.split("_")]
+        )
+        plugin_arguments_model_name = f"{plugin_type_camel_case}ArgumentsModel"
+        plugin_arguments_model = globals().get(plugin_arguments_model_name)
+        if plugin_arguments_model is None:
+            raise ValueError(
+                f"\n\n\tThe plugin type argument class"
+                f"{plugin_arguments_model_name} is not defined\n\n"
+            )
+        plugin_arguments_model(**values.get("arguments", {}))
+
+        return values
+
+
+class ProductStepModel(BaseModel):
+    """Validate and process a sequence of steps with their data."""
+
+    definition: ProductStepDefinitionModel = Field(..., description="Sequence of steps")
+
+    @model_validator(mode="before")
+    def _plugin_name_validator(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Validate user input for the plugin type.
+
+        Parameters
+        ----------
+        values : dict
+            The step data containing step name (plugin type) and its content.
+
+        Returns
+        -------
+        dict
+            The processed step data after step name validation
+
+        Raises
+        ------
+        ValueError
+            if the user input step name is not in the valid_types list
+            if the user input step name and type (if provided) are not same
+        """
+        if not values:
+            raise ValueError("Empty : Step data cannot be empty.")
+
+        valid_types = get_plugin_types()
+
+        # extract step name and step data
+        plugin_type, step_data = next(iter(values.items()))
+
+        # raise error if the step name (plugin type) is not valid
+        if plugin_type not in valid_types:
+            raise ValueError(
+                f"\n\ninvalid step name : {plugin_type}.\n\t"
+                f"Must be one of {valid_types}\n\n"
+            )
+
+        # extract value for type field (if provided) otherwise
+        # add the key value for step name
+        if "type" not in step_data:
+            step_data = {"type": plugin_type.lower(), **step_data}
+            # print("plugin_type ", plugin_type)
+
+        # ensure 'type' field matches step name
+        if step_data["type"] != plugin_type:
+            raise ValueError(
+                f"\n\nstep name : '{plugin_type}'"
+                f"and type : '{step_data['type']}' mismatch. "
+                f"Check your product definition\n\n"
+            )
+
+        return {"definition": step_data}
+
+
+class ProductSpecModel(BaseModel):
+    """The specification for a product."""
+
+    # list of steps
+    steps: List[ProductStepModel] = Field(
+        ..., description="Steps to produce the product."
+    )
+
+
+class ProductPluginModel(PluginModel, BaseModel):
```

Comment:
I think we can remove `BaseModel` here since `PluginModel` inherits from `PrettyBaseModel` which inherits from `BaseModel`.

Reply:
Thank you, @jsolbrig for catching this. BaseModel is removed and the results along with tests are verified as well.